modcc/cudaprinter.cpp

```diff
@@ -164,6 +167,36 @@ CUDAPrinter::CUDAPrinter(Module &m, bool o)
             }
         }
     }
+
+    // print the write_back kernel
+    if(module_->write_backs().size()) {
+        buffer().add_line("__global__");
+        buffer().add_line("void write_back_"+module_name_+"("+pack_name()+" params_) {");
+        buffer().increase_indentation();
+        buffer().add_line("using value_type = arb::fvm_value_type;");
+
+        buffer().add_line("auto tid_ = threadIdx.x + blockDim.x*blockIdx.x;");
+        buffer().add_line("auto const n_ = params_.n_;");
+        buffer().add_line("if(tid_<n_) {");
+        buffer().increase_indentation();
+
+        auto ion_prefix = [](ionKind k) {
+            return k==ionKind::Ca? "ca": k==ionKind::Na? "na": "k";
+        };
+
+        for (auto& w: module_->write_backs()) {
+            auto& src = w.source_name;
+            auto& tgt = w.target_name;
+
+            auto idx = src + "_idx_";
+            buffer().add_line("auto "+idx+" = params_.ion_"+ion_prefix(w.ion_kind)+"_idx_[tid_];");
+            buffer().add_line("// 1/10 magic number due to unit normalisation");
+            buffer().add_line("params_."+tgt+"["+idx+"] = value_type(0.1)*params_.weights_[tid_]*params_."+src+"[tid_];");
```

Comment:
On one hand, the target of an ion write back is a (backend-specific) function of the corresponding variable, and on the other the `CPrinter` and `CUDAPrinter` handling of the `target_name` field from the `Module::write_backs()` call differs, and looks a bit magic without knowing the whole path from ION block parsing down to this bit of code.

I think it would simplify both the logic and the understanding of the code if the printers just used a list of ion-write variables. This information is already available from the `IonDep` data.

Reply:
This is going to get ripped up when generic ion species are introduced, so I am not inclined to really refactor fully now.

Reply:
I'll probably hack on it in the mechanism PR of doom.